core/src/main/java/org/jboss/jandex/StackedIndex.java

```diff
@@ -0,0 +1,326 @@
+package org.jboss.jandex;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * A stack of {@linkplain IndexView indexes} with overlay semantics. Overlaying is done on class
+ * granularity. That is, if a class is present in multiple indexes on the stack, only the top-most
+ * occurrence is considered; the other variants of the class present below on the stack are ignored.
+ *
+ * @since 3.2.0
+ */
+public final class StackedIndex implements IndexView {
+    // note that the top-most index comes first (reverse order compared to the `create()` methods)
+    private final IndexView[] stack;
+
+    private StackedIndex(IndexView[] stack) {
+        this.stack = stack;
+    }
+
+    /**
+     * Creates a stacked index from given {@code indexes}. The first element of the list is the bottom-most index
```

Comment:
This means that when you build the stack you need to add the indexes with "lower priority" first which is IMO not exactly intuitive and practical. I wonder if we should offer some more flexible way of building the stack.

Reply:
I actually had a `push(IndexView)` method implemented (that would return a new stacked index with the given argument on top of the stack), but I felt like a solution in search of a problem, so I ultimately removed it. It should be easy to re-add.

Reply:
Added `StackedIndex.pushIndex()`.